### ISSUE TITLE
[RELEASE REGRESSIONS] - Library Access, Hand Reveal, Sideboard Access.

### ISSUE THREAD

**Issue by woasha** [Cockatrice/Cockatrice#588]
Using the F3 function, or menu selection to view entire library causes client to crash. 

Example using menu option and then shortcut.http://screencast.com/t/8pVfXDYXZO

Platform: OSX
Version: Jan,23rd, 2015. 

Update: Regression appears to also effect accessing the sideboard. Behaviour is same when selecting view sideboard, or using keyboard shortcuts to access it. 

Update: When revealing your entire hand it also causes a full client crash. http://screencast.com/t/ohelseUXX1a

**Comment by ZeldaZach** [comment]
Couldn't reproduct running version 712f12a

**Comment by woasha** [comment]
The Version number on the OS X download version is b06b818. Has the OS X platform download source been updated also?

**Comment by Daenyth** [comment]
I can't reproduce using my compiled client from that commit with qt5

**Comment by woasha** [comment]
Still persisting for me after two reinstalls (prior and after a reboot).Version hash from native OS X latest install still yields b06b818. Tried the release before on Jan22nd, and it exhibited the same issues. I went further back to version# 7476667 on Jan. 5th and that one actually functions correctly prior to the latest releases which is indicative of an actual, functional regression. 

I've noticed also, that others are starting to voice functional concerns for crashes pertaining to a recent release.

**Comment by ZeldaZach** [comment]
@Daenyth I think it might be a QT4 issue with a newer function we added in the past month

**Comment by woogerboy21** [comment]
And it is only isolated to OSX correct?

**Comment by MassCraziness** [comment]
@woogerboy21 No issues with it on Windows.

**Comment by wildnacatl** [comment]
Just installed for the first time today.  I right clicked on my library and said View Library and my client crashes.  

OSX 10.9.5
Cockatrice version b06b818

**Comment by ZeldaZach** [comment]
This seems like to be a persistent issue among the build. @woogerboy21 You're building qt4 on 10.6 still?

**Comment by wildnacatl** [comment]
I just tried viewing my graveyard and it crashed, as well.  Also crashes when I use the F4 shortcut. Not sure if it's the same issue.

**Comment by Daenyth** [comment]
I suspect this problem is with the qt4 version of the pile sorting feature. Cc @poixen

**Comment by Daenyth** [comment]
Either that or the fact that it's compiled on 10.6 - works for me with mavericks qt5

**Comment by anoda9** [comment]
On mac os 10.6.8, cockatrice version b06b818: I double click to view a graveyard and the program crashes.

**Comment by ZeldaZach** [comment]
I think the issue is qt4...

**Comment by tooomm** [comment]
Time to move on maybe? Isn't 10.6 like Windows XP? :)

Well, even Dropbox drops support for 10.6 soon... (starting 18th of May)

**Comment by Daenyth** [comment]
If it is qt4 we may still need to consider dropping support for 10.6 - maintaining the pile split option for qt5 only may be annoying

**Comment by ctrlaltca** [comment]
From the screencast, the crash is always at:
```
ZoneViewZone::setPileViewPositions(int, CardList&, int&, int&) + 135
```
Disassembling the method:
```
Dump of assembler code for function _ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_:
0x00000001000c0e70 <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+0>:	push   %rbp
0x00000001000c0e71 <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+1>:	mov    %rsp,%rbp
0x00000001000c0e74 <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+4>:	sub    $0xa0,%rsp
0x00000001000c0e7b <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+11>:	mov    %rdi,-0x8(%rbp)
0x00000001000c0e7f <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+15>:	mov    %esi,-0xc(%rbp)
0x00000001000c0e82 <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+18>:	mov    %rdx,-0x18(%rbp)
0x00000001000c0e86 <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+22>:	mov    %rcx,-0x20(%rbp)
0x00000001000c0e8a <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+26>:	mov    %r8,-0x28(%rbp)
0x00000001000c0e8e <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+30>:	movl   $0x0,-0x2c(%rbp)
0x00000001000c0e95 <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+37>:	movb   $0x1,-0x2d(%rbp)
0x00000001000c0e99 <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+41>:	movl   $0x0,-0x34(%rbp)
0x00000001000c0ea0 <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+48>:	mov    -0x34(%rbp),%eax
0x00000001000c0ea3 <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+51>:	cmp    -0xc(%rbp),%eax
0x00000001000c0ea6 <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+54>:	jge    0x1000c1098 <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+552>
0x00000001000c0eac <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+60>:	mov    -0x18(%rbp),%rax
0x00000001000c0eb0 <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+64>:	mov    -0x34(%rbp),%esi
0x00000001000c0eb3 <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+67>:	mov    %rax,%rdi
0x00000001000c0eb6 <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+70>:	callq  0x1003902e4 <dyld_stub__ZNK5QListIP8CardItemE2atEi>
0x00000001000c0ebb <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+75>:	mov    (%rax),%rax
0x00000001000c0ebe <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+78>:	mov    %rax,-0x40(%rbp)
0x00000001000c0ec2 <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+82>:	mov    -0x40(%rbp),%rax
0x00000001000c0ec6 <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+86>:	mov    %rax,%rdi
0x00000001000c0ec9 <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+89>:	callq  0x100390044 <dyld_stub__ZNK16AbstractCardItem7getInfoEv>
0x00000001000c0ece <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+94>:	lea    -0x48(%rbp),%rdi
0x00000001000c0ed2 <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+98>:	mov    %rax,%rsi
0x00000001000c0ed5 <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+101>:	callq  0x1000813f0 <_ZNK8CardInfo15getMainCardTypeEv>
0x00000001000c0eda <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+106>:	cmpl   $0x0,-0x34(%rbp)
0x00000001000c0ee1 <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+113>:	je     0x1000c0fe0 <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+368>
0x00000001000c0ee7 <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+119>:	mov    -0x18(%rbp),%rdi
0x00000001000c0eeb <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+123>:	mov    -0x34(%rbp),%eax
0x00000001000c0eee <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+126>:	dec    %eax
0x00000001000c0ef0 <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+128>:	mov    %eax,%esi
0x00000001000c0ef2 <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+130>:	callq  0x1003902e4 <dyld_stub__ZNK5QListIP8CardItemE2atEi>
0x00000001000c0ef7 <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+135>:	mov    %rax,-0x78(%rbp)
0x00000001000c0efb <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+139>:	jmpq   0x1000c0f00 <_ZN12ZoneViewZone20setPileViewPositionsEiR8CardListRiS2_+144>
```
The crash happens at line 138, because at(i-1) is not returning a valid card:
```
cardTypeMatch = cardType.compare(cardsToDisplay.at(i-1)->getInfo()->getMainCardType()) == 0 ? true : false;
```
I suppose commit e4e3a1d could be the culprit.

**Comment by ctrlaltca** [comment]
By the way i'd try refactoring that line, keeping in a variable the value of cardType from the previous loop iteration, like this:
```c++
void ZoneViewZone::setPileViewPositions(int cardCount, CardList &cardsToDisplay, int &typeColumn, int &longestRow){
    int typeRow = 0;
    bool cardTypeMatch = true;
    CardItem *c;
    QString cardType, lastCardType;
    for (int i = 0; i < cardCount; i++) {
        c = cardsToDisplay.at(i);
        cardType = c->getInfo()->getMainCardType();

        if(cardType == lastCardType)
        {
            // add below current card
            typeRow++;   
        } else {
            // create new column
            typeColumn++;
            longestRow = qMax(typeRow, longestRow);
            typeRow = 0;
            lastCardType = cardType;
        }
                
        qreal x = typeColumn * CARD_WIDTH;
        qreal y = typeRow * CARD_HEIGHT / 3;
        c->setPos(x + 5, y + 5);
        c->setRealZValue(i);
    }
}
```

**Comment by Daenyth** [comment]
Some testing results:

qt5 on mavericks, self compiled - works
qt4 on mavericks, self compiled - works
qt4 on mavericks, wooger weekly build - crashes

**Comment by poixen** [comment]
@ctrlaltca I like some of your impl and will work on it tonight. I see you moved the CardItem* outside the loop, what was the reason for this? As the ref will be changed at the start of each loop. Maybe Im not seeing something ...

**Comment by ctrlaltca** [comment]
It could make a difference if the variable was not a pointer/primitive type/POD, so as a "best practice" i use to always move any variable declaration outside loops.
In this case, moving the variable declaration outside the loop makes no benefit, since the var is a pointer and the compiler is smart enough to understand it needs to be allocated on the stack only once.

**Comment by Daenyth** [comment]
Fixed in #607

**Comment by woasha** [comment]
Smoke tested - So far so good. +1.
